clash-lib/src/Clash/Core/Evaluator/Semantics.hs

```diff
@@ -0,0 +1,372 @@
+-- evaluated to obtain a result.
+--
+evaluateDataWith
+  :: (Term -> State Env Value)
+  -> DataCon
+  -> State Env Value
+evaluateDataWith eval dc
+  | isFullyApplied (dcType dc) [] = pure (VData dc [])
+  | otherwise = etaExpand (Data dc) >>= eval
+
+-- | Default implementation for evaluating primitive operations. If the primop
+-- is not fully applied, then it is eta-expanded and its eta-expanded form is
+-- delta-reduced to obtain a result.
+--
+evaluatePrimWith
+  :: (Term -> State Env Value)
+  -> (PrimInfo -> [Either Value Type] -> State Env Value)
+  -> PrimInfo
+  -> State Env Value
+evaluatePrimWith eval evalPrim p
+  | isFullyApplied (primType p) [] = evalPrim p []
+  | otherwise = etaExpand (Prim p) >>= eval
+
+-- | Default implementation for evaluating lambdas. As a term with a lambda
+-- at the head is already in WHNF, this simply returns the term under the
+-- lambda with the current environment.
+--
+evaluateLam :: Id -> Term -> State Env Value
+evaluateLam i x = State.gets (VLam i x)
+
+-- | Default implementation for evaluating type lambdas. As a term with a type
+-- lambda at the head is already in WHNF, this simply returns the term under
+-- the type lambda with the current environment.
+--
+evaluateTyLam :: TyVar -> Term -> State Env Value
+evaluateTyLam i x = State.gets (VTyLam i x)
+
+-- | Default implementation for evalating a term / type application. This
+-- checks if the arguments are applied to a data constructor or primitive, and
+-- creates a value if it is fully applied. If the arguments are applied to a
+-- function, the first argument is applied.
+--
+evaluateAppWith
+  :: (Term -> State Env Value)
+  -> (PrimInfo -> [Either Value Type] -> State Env Value)
+  -> (Value -> Either Value Type -> State Env Value)
+  -> Term
+  -> Either Term Type
+  -> State Env Value
+evaluateAppWith eval evalPrim apply x y
+  | Data dc <- f
+  = if isFullyApplied (dcType dc) args
+       then VData dc <$> evalArgs args
+       else etaExpand term >>= eval
+
+  | Prim p <- f
+  , nArgs  <- length . fst $ splitFunForallTy (primType p)
+  = case compare (length args) nArgs of
+      LT -> etaExpand term >>= eval
+      EQ -> evalArgs args >>= evalPrim p
+      GT -> do
+        (pArgs, rArgs) <- splitAt nArgs <$> evalArgs args
+        primRes <- evalPrim p pArgs
+        foldM apply primRes rArgs
+
+  | otherwise
+  = do
+       evalX <- eval x
+       evalY <- evalArg y
+       apply evalX evalY
+ where
+  term      = either (App x) (TyApp x) y
+  (f, args) = collectArgs term
+  evalArg   = bitraverse eval pure
+  evalArgs  = traverse evalArg
+
+-- | Default implementation for evaluating a letrec expression. This adds all
+-- bindings to the heap without eagerly evaluating them, then evaluates the
+-- body of the expression.
+--
+-- Bindings are not evaluated eagerly, as they may not be needed depending how
+-- the body of the expression evaluates (i.e. bindings may be in an elided
```

Comment:
Elide, as in omitted. Eliminate isn't _quite_ the word I'd want to use for this because it means something in terms of reduction (whereas this is more of a pruning)

Reply:
I had to look up elide.
I could guess based on context, but it wasn't immediately. 
Any of omitted,pruned,removed,eliminated seem much clearer to me.

Reply:
Fair enough, in that case I'll change things to say pruned. It has the strongest "I'm cutting this out of the term" connotation

Reply:
If *elide* captures the intention and nuance better than any of its "synonyms", I think I'd rather people grab a dictionary to understand the comments, than to change the comment.